test/jdk/java/net/Socket/CloseOnFailureTest.java

```diff
@@ -0,0 +1,272 @@
+
+        private final Exception connectError;
+
+        private MockSocketImpl(Exception bindError, Exception connectError) {
+            this.bindError = bindError;
+            this.connectError = connectError;
+        }
+
+        @Override
+        protected void create(boolean stream) {}
+
+        @Override
+        protected void bind(InetAddress host, int port) throws IOException {
+            throwIfPresent(bindError);
+        }
+
+        @Override
+        protected void connect(SocketAddress address, int timeoutMillis) throws IOException {
+            throwIfPresent(connectError);
+        }
+
+        private void throwIfPresent(Exception connectError) throws IOException {
+            if (connectError != null) {
+                switch (connectError) {
+                    case IOException error -> throw error;
+                    case RuntimeException error -> throw error;
+                    default -> throw new IllegalStateException(
+                            "unknown connectError type: " + connectError.getClass().getCanonicalName());
+                }
+            }
+        }
+
+        @Override
+        protected void close() {
+            closeInvocationCounter.incrementAndGet();
+        }
+
+    }
+
+    @ParameterizedTest
+    @MethodSource("ctorShouldCloseOnFailuresTestCases")
+    @SuppressWarnings("resource")
+    void ctorShouldCloseOnFailures(TestCase testCase) throws Throwable {
+
+        // Create a socket using the mock `SocketImpl` configured to fail
+        withSocketImplFactory(() -> testCase.socketImpl, () -> {
+
+            // Trigger the failure
+            Exception error = assertThrows(Exception.class, () -> {
+                // Address and port are mostly ineffective.
+                // They just need to be _valid enough_ to reach to the point where both `SocketImpl#bind()` and `SocketImpl#connect()` are invoked.
+                // Failure will be triggered by the injected `SocketImpl`.
+                InetAddress serverAddress = InetAddress.getLoopbackAddress();
+                int deadServerPort = 0xDEAD;
+                new Socket(serverAddress, deadServerPort, null, 0);
+            });
+
+            // Run verifications
+            testCase.caughtErrorVerifier.accept(error);
+            testCase.socketImplVerifier.run();
+
+        });
+
+    }
+
+    static List<TestCase> ctorShouldCloseOnFailuresTestCases() {
+        return List.of(
+                TestCase.BindFailureFactory.iOExceptionTestCase(),
+                TestCase.ConnectFailureFactory.iOExceptionTestCase(),
+                TestCase.ConnectFailureFactory.illegalArgumentExceptionTestCase(1));
+    }
+
+    @Test
+    void connectShouldCloseOnUnresolvedAddress() throws IOException {
+        MockSocketImpl socketImpl = new MockSocketImpl(null, null);
+        try (Socket socket = new Socket(socketImpl) {}) {
+            InetSocketAddress address = InetSocketAddress.createUnresolved("no.such.host", 0xBEEF);
```

Comment:
similar to DEADSERVERPORT

specifying a hex value, albeit a joke value, it is still a valid ephemeral port, and distracts from the main invocation InetAddress.createUnresolved, so consider a constant declaration for the port

private static final int NOSUCHHOSTPORT = 0xBEEF;

then the invocation is
InetAddress.createUnresolved("no.such.host", NOSUCHHOSTPORT);

presenting an unabiquous statement that the InetAddress not meant to be reachable !

Reply:
Reused the `DEAD_SERVER_PORT` constant here in 35d680123f83b97b2a62fe99969aa0c18a75e6b1.